x/wasm/IBC.md

```diff
@@ -0,0 +1,281 @@
+* If a *Channel* is being established with a registered `wasm-xyz` port,
+  the `x/wasm.Keeper` will handle this and call into the appropriate
+  contract to determine supported protocol versions during the
+  [`ChanOpenTry` and `ChanOpenAck` phases](https://docs.cosmos.network/master/ibc/overview.html#channels).
+  (See [Channel Handshake Version Negotiation](https://docs.cosmos.network/master/ibc/custom.html#channel-handshake-version-negotiation))
+* Both the *Port* and the *Channel* are fully owned by one contract.
+* `x/wasm` will allow both *ORDERED* and *UNORDERED* channel and pass that mode
+  down to the contract in `OnChanOpenTry`, so the contract can decide if it accepts
+  the mode. We will recommend the contract developers stick with *ORDERED* channels
+  unless they can reason about async packet timing.
+* When sending a packet, the CosmWasm contract must specify the *ChannelID*
+  and remote *PortID*, which is generally provided by the external client
+  which understands which channels it wishes to communicate over.
+* When receiving a Packet (or Ack or Error), the contracts receives the
+  *ChannelID* as well as remote *PortID* (and *ClientID*???) that is
+  communicating.
+* When receiving an Ack or Error packet, the contract also receives the
+  original packet that it set on Send 
+  
+  
+## Workflow
+
+Establishing *Clients* and *Connections* is out of the scope of this
+module and must be created by the same means as for `ibc-transfer`
+(via the cli or otherwise). `x/wasm` will bind a unique *Port* for each
+"IBC Enabled" contract.
+
+An "IBC Enabled" contract may dispatch the following messages not available
+to other contracts:
+
+* `IBCSendMsg` - this sends an IBC packet over an established channel. 
+* `IBCOpenChannel` - given a ConnectionID and a remote Port, this will request
+  to open a new channel with the given chain
+* `IBCCloseChannel` - given an existing channelID bound to this contract's Port,
+  initiate the closing sequence and reject all pending packets.
+
+They are returned from `handle` just like any other `CosmosMsg`
+(For mocks, we will trigger this externally, later only valid contract addresses 
+should be able to do so).
+
+An "IBC Enabled" contract must support the following callbacks from the runtime
+(we will likely multiplex many over one wasm export, as we do with handle, but these 
+are the different calls we must support):
+
+Packet Lifecycle:
+
+* `IBCRecvPacket` - when another chain sends a packet destined to
+  an IBCEnabled contract, this will be routed to the proper contract
+  and call an exposed function for this purpose.
+* `IBCPacketAck` - The original sender of `IBCSendMsg` will
+  get this callback eventually if the message was
+  processed on the other chain (this may be either a success or an error,
+  but comes from the app-level protocol, not the IBC protocol).
+* `IBCPacketDropped` - The original sender of `IBCSendMsg` will 
```

Comment:
You can name it whatever you like, but this is all called a "timeout" in IBC (including channel closure)

Reply:
Okay, I wasn't sure about all the different cases (timeout, channel closed, client no longer valid...)

I will use the standard naming.